core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -141,4 +136,41 @@ private static String stripTrailingSlash(String path) {
     }
     return result;
   }
+
+  /**
+   * Get the data file location. For the {@link DefaultLocationProvider}, the priority level are
+   * "write.data.path" -> "write.folder-storage.path" -> "table-location/data".
+   * For the {@link ObjectStoreLocationProvider}, the priority level are
+   * "write.data.path" -> "write.object-storage.path" -> "write.folder-storage.path" -> "table-location/data".
+   */
+  private static String dataLocation(Map<String, String> properties, String tableLocation, boolean isObjectStore) {
+    String dataLocation = properties.get(TableProperties.WRITE_DATA_LOCATION);
+    if (dataLocation == null) {
+      dataLocation = deprecatedDataLocation(properties, isObjectStore);
+      if (dataLocation == null) {
+        dataLocation = String.format("%s/data", tableLocation);
+      }
+    }
+    return dataLocation;
+  }
+
+  private static String deprecatedDataLocation(Map<String, String> properties, boolean isObjectStore) {
+    String deprecatedProperty = isObjectStore ?
+        TableProperties.OBJECT_STORE_PATH : TableProperties.WRITE_FOLDER_STORAGE_LOCATION;
+
+    String dataLocation = properties.get(deprecatedProperty);
+
+    final String warnMsg = "Table property {} is deprecated, please use " + TableProperties.WRITE_DATA_LOCATION +
+        " instead.";
+    if (dataLocation != null) {
+      LOG.warn(warnMsg, deprecatedProperty);
+    } else if (deprecatedProperty.equals(TableProperties.OBJECT_STORE_PATH)) {
+      dataLocation = properties.get(TableProperties.WRITE_FOLDER_STORAGE_LOCATION);
+      if (dataLocation != null) {
+        LOG.warn(warnMsg, TableProperties.WRITE_FOLDER_STORAGE_LOCATION);
+      }
+    }
+
+    return dataLocation;
```

Comment:
This looks way too complicated to be worth it. Let's just duplicate the logic and minimize the changes in each location provider.

Reply:
Made the change.